packages/apollo/src/commands/service/__tests__/check.test.ts

```diff
@@ -1,22 +1,328 @@
+
+/**
+ * The original `console.log` being mocked.
+ *
+ * We save it so we can restore it after a test.
+ */
+let mockedConsoleLogOriginal: Console["log"] | null = null;
+
+/**
+ * Array of intercepted console values.
+ */
+let mockedConsoleLogValues: string[] | null = null;
+
+/**
+ * Mock and capture `console.log` and `stdout.write`s. Return them in that order as a single string.
+ *
+ * This will emulate what the output of running the CLI would look like.
+ *
+ * Call `uncaptureApplicationOutput` to reverse the effects of this function.
+ */
+function captureApplicationOutput() {
+  mockedConsoleLogOriginal = console["log"];
+  mockedConsoleLogValues = [];
+  console["log"] = jest.fn((...items) => {
+    if (!mockedConsoleLogValues) {
+      throw new Error(
+        "mockedConsoleLogValues is not prepared but we're still capturing console.log. This means there's a bug somewhere."
+      );
+    }
+
+    mockedConsoleLogValues.push(items.join(" "));
+  });
+
+  stdout.start();
+}
+
+/**
+ * Reverse mocking of `console.log` and `stdout.write`. If they weren't mocked to begin with, this will do
+ * nothing and return null.
+ */
+function uncaptureApplicationOutput(): string | null {
+  // These will be `null` if we haven't mocked `console.log`.
+  if (!mockedConsoleLogOriginal || !mockedConsoleLogValues) {
+    return null;
+  }
+
+  const result = mockedConsoleLogValues.concat(stdout.output).join("\n");
+  mockedConsoleLogValues = null;
+
+  // Restore `console.log`
+  console["log"] = mockedConsoleLogOriginal;
+
+  // Stop capturing `stdout`.
+  stdout.stop();
+
+  return result;
+}
+
+/**
+ * Convert a schema SDL to an introspection query result.
+ *
+ * @see https://blog.apollographql.com/three-ways-to-represent-your-graphql-schema-a41f4175100d
+ *
+ * @param schemaSdl string Schema in SDL form
+ */
+function sdlToIntrospectionQueryResult(schemaSdl: string) {
+  return graphql.graphqlSync(
+    graphql.buildSchema(schemaSdl),
+    graphql.introspectionQuery
+  ).data;
+}
+
+/**
+ * Use `nock` to mock an `IntrospectionQuery`
+ *
+ * @param url string Root of the URL to mock; `/graphql` will automatically be appended
+ * @param sdl SDL of the schema to mock
+ */
+function mockIntrospectionQuery() {
+  nock(localURL, { encodedQueryParams: true })
+    .post("/graphql", request => request.operationName === "IntrospectionQuery")
+    .reply(200, {
+      // The SDL doesn't actually get used because we'll be simulating network responses regardless of input,
+      // so we just use a fake SDL.
+      data: sdlToIntrospectionQueryResult(`type Query { me: ID }`)
+    });
+}
+
+/**
+ * Mock network requests for a non-federated schema check that produces errors.
+ */
+function mockNonFederatedFailure() {
+  mockIntrospectionQuery();
+
+  nock("https://engine-staging-graphql.apollographql.com:443", {
+    encodedQueryParams: true
+  })
+    .post("/api/graphql", () => true)
+    .reply(200, {
+      data: {
+        service: {
```

Comment:
Can we use our generated types to type this data? That way we know to update it and test the right thing when the schema / types change!

Reply:
Some of it yes, some of it no. I started this and it ended up being a pain because we have to add all the `__typename`s and we can't use string literals for enum values. If it's ok, I'd rather _not_ do this for now 🤷‍♂

Reply:
That's ok with me!